### PR TITLE
Check for converter errors

### PR DESCRIPTION
**Pull request by kunsingh**
This change is [Reviewable](https://reviewable.io/reviews/canoo/dolphin-platform/778)

### CODE REVIEW

File: platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/DurationConverterFactory.java [hunk 1]
```diff
@@ -36,6 +39,11 @@ public boolean supportsType(Class<?> cls) {
         return Duration.class.isAssignableFrom(cls);
     }
 
+    @Override
+    public List<Class> getSupportedTypes() {
+        return new ArrayList<>(Arrays.asList(Duration.class));
```
Comment (hendrikebbers) [inline platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/DurationConverterFactory.java]: Collections.singletonList(Duration.class);

File: platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/LocalDateConverterFactory.java [hunk 2]
```diff
@@ -45,6 +43,11 @@ public boolean supportsType(Class<?> cls) {
         return LocalDate.class.isAssignableFrom(cls);
     }
 
+    @Override
+    public List<Class> getSupportedTypes() {
+        return new ArrayList<>(Arrays.asList(LocalDate.class));
```
Comment (hendrikebbers) [inline platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/LocalDateConverterFactory.java]: Collections.singletonList

File: platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/LocalDateTimeConverterFactory.java [hunk 3]
```diff
@@ -45,6 +43,11 @@ public boolean supportsType(Class<?> cls) {
         return LocalDateTime.class.isAssignableFrom(cls);
     }
 
+    @Override
+    public List<Class> getSupportedTypes() {
+        return new ArrayList<>(Arrays.asList(LocalDateTime.class));
```
Comment (hendrikebbers) [inline platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/LocalDateTimeConverterFactory.java]: Collections.singletonList

File: platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/PeriodConverterFactory.java [hunk 4]
```diff
@@ -36,6 +39,11 @@ public boolean supportsType(Class<?> cls) {
         return Period.class.isAssignableFrom(cls);
     }
 
+    @Override
+    public List<Class> getSupportedTypes() {
+        return new ArrayList<>(Arrays.asList(Period.class));
```
Comment (hendrikebbers) [inline platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/PeriodConverterFactory.java]: Collections.singletonList

File: platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/ZonedDateTimeConverterFactory.java [hunk 5]
```diff
@@ -44,6 +42,11 @@ public boolean supportsType(Class<?> cls) {
         return ZonedDateTime.class.isAssignableFrom(cls);
     }
 
+    @Override
+    public List<Class> getSupportedTypes() {
+        return new ArrayList<>(Arrays.asList(ZonedDateTime.class));
```
Comment (hendrikebbers) [inline platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/ZonedDateTimeConverterFactory.java]: Collections.singletonList

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java [hunk 6]
```diff
@@ -47,9 +47,30 @@ public Converters(final BeanRepository beanRepository) {
...
-            factory.init(beanRepository);
-            converterFactories.add(factory);
+            if(!isTypeAlreadyAdded(factory) && !isClassAlreadyAdded(factory)) {
+                factory.init(beanRepository);
+                converterFactories.add(factory);
+            }
+        }
+    }
+
+    private boolean isTypeAlreadyAdded(ConverterFactory converterFactory) {
```
Comment (hendrikebbers) [inline platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java]: final

Comment (hendrikebbers) [inline platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java]: Assert.notNull

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java [hunk 7]
```diff
@@ -47,9 +47,30 @@ public Converters(final BeanRepository beanRepository) {
...
+        for(ConverterFactory factory : converterFactories){
+            if(factory.getTypeIdentifier() == converterFactory.getTypeIdentifier()){
+                return true;
+            }
+        }
+        return false;
+    }
+
+
+    private boolean isClassAlreadyAdded(ConverterFactory converterFactory) {
```
Comment (hendrikebbers) [inline platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java]: This should check factory. getSupportedTypes() and not the class of the factory ;)

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java [hunk 8]
```diff
@@ -47,9 +47,30 @@ public Converters(final BeanRepository beanRepository) {
         while (iterator.hasNext()) {
             ConverterFactory factory = iterator.next();
             LOG.trace("Found converter factory {} with type identifier {}", factory.getClass(), factory.getTypeIdentifier());
-            factory.init(beanRepository);
-            converterFactories.add(factory);
+            if(!isTypeAlreadyAdded(factory) && !isClassAlreadyAdded(factory)) {
+                factory.init(beanRepository);
+                converterFactories.add(factory);
+            }
```
Comment (hendrikebbers) [inline platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java]: Here an exception should be thrown if we have a conflict

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/converters/BigDecimalConverterFactory.java [hunk 9]
```diff
@@ -54,6 +57,11 @@ public boolean supportsType(Class<?> cls) {
         return BigDecimal.class.equals(cls);
     }
 
+    @Override
+    public List<Class> getSupportedTypes() {
+        return new ArrayList<>(Arrays.asList(BigDecimal.class));
```
Comment (hendrikebbers) [inline platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/converters/BigDecimalConverterFactory.java]: Collections.singletonList

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/converters/BigIntegerConverterFactory.java [hunk 10]
```diff
@@ -54,6 +57,11 @@ public boolean supportsType(Class<?> cls) {
         return BigInteger.class.equals(cls);
     }
 
+    @Override
+    public List<Class> getSupportedTypes() {
+        return new ArrayList<>(Arrays.asList(BigInteger.class));
```
Comment (hendrikebbers) [inline platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/converters/BigIntegerConverterFactory.java]: Collections.singletonList

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/converters/BooleanConverterFactory.java [hunk 11]
```diff
@@ -43,6 +47,11 @@ public boolean supportsType(Class<?> cls) {
         return boolean.class.equals(cls) || Boolean.class.equals(cls);
     }
 
+    @Override
+    public List<Class> getSupportedTypes() {
+        return new ArrayList<>(Arrays.asList(boolean.class, Boolean.class));
```
Comment (hendrikebbers) [inline platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/converters/BooleanConverterFactory.java]: return Arrays.asList(boolean.class, Boolean.class)

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/converters/ByteConverterFactory.java [hunk 12]
```diff
@@ -44,6 +48,11 @@ public boolean supportsType(Class<?> cls) {
         return byte.class.equals(cls) || Byte.class.equals(cls);
     }
 
+    @Override
+    public List<Class> getSupportedTypes() {
+        return new ArrayList<>(Arrays.asList(byte.class, Byte.class));
```
Comment (hendrikebbers) [inline platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/converters/ByteConverterFactory.java]: return Arrays.asList(....)

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/platform/remoting/spi/converter/ConverterFactory.java [hunk 13]
```diff
@@ -46,6 +48,12 @@
      */
     boolean supportsType(Class<?> cls);
 
+    /**
+     * This method will be called to get all the supported converter types
```
Comment (hendrikebbers) [inline platform/dolphin-platform-remoting-common/src/main/java/com/canoo/platform/remoting/spi/converter/ConverterFactory.java]: ...all supported types for conversion

**Comment by hendrikebbers** [comment]
fix for https://github.com/canoo/dolphin-platform/issues/756

File: platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/DurationConverterFactory.java (hunk 1 above)
Comment (kunsingh) [reply platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/DurationConverterFactory.java]: Done.

File: platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/LocalDateConverterFactory.java (hunk 2 above)
Comment (kunsingh) [reply platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/LocalDateConverterFactory.java]: Done.

File: platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/LocalDateTimeConverterFactory.java (hunk 3 above)
Comment (kunsingh) [reply platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/LocalDateTimeConverterFactory.java]: Done.

File: platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/PeriodConverterFactory.java (hunk 4 above)
Comment (kunsingh) [reply platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/PeriodConverterFactory.java]: Done.

File: platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/ZonedDateTimeConverterFactory.java (hunk 5 above)
Comment (kunsingh) [reply platform/dolphin-platform-date-time-converter/src/main/java/com/canoo/dolphin/converters/ZonedDateTimeConverterFactory.java]: Done.

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java (hunk 8 above)
Comment (kunsingh) [reply platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java]: Done. Pls check if the exception is right.

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java (hunk 6 above)
Comment (kunsingh) [reply platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java]: Done.

Comment (kunsingh) [reply platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java]: Done.

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java (hunk 7 above)
Comment (kunsingh) [reply platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java]: Done.

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/converters/BigDecimalConverterFactory.java (hunk 9 above)
Comment (kunsingh) [reply platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/converters/BigDecimalConverterFactory.java]: Done.

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/converters/BigIntegerConverterFactory.java (hunk 10 above)
Comment (kunsingh) [reply platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/converters/BigIntegerConverterFactory.java]: Done.

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/converters/BooleanConverterFactory.java (hunk 11 above)
Comment (kunsingh) [reply platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/converters/BooleanConverterFactory.java]: Done.

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/converters/ByteConverterFactory.java (hunk 12 above)
Comment (kunsingh) [reply platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/converters/ByteConverterFactory.java]: Done.

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/platform/remoting/spi/converter/ConverterFactory.java (hunk 13 above)
Comment (kunsingh) [reply platform/dolphin-platform-remoting-common/src/main/java/com/canoo/platform/remoting/spi/converter/ConverterFactory.java]: Done.

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java [hunk 14]
```diff
@@ -47,20 +46,48 @@ public Converters(final BeanRepository beanRepository) {
...
+        }
+        return false;
+    }
+
+
+    private boolean isClassAlreadyAdded(final ConverterFactory converterFactory) {
+        Assert.requireNonNull(converterFactory, "converterFactory");
+        for(ConverterFactory factory : converterFactories){
+            List<Class> list = factory.getSupportedTypes();
```
Comment (hendrikebbers) [inline platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java]: final

Comment (hendrikebbers) [reply platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java]: Should be refactored in:
return factory.getSupportedTypes().filter(c -> ....).findAny().map(c -> true).orElse(false);

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java [hunk 15]
```diff
@@ -47,20 +46,48 @@ public Converters(final BeanRepository beanRepository) {
...
+        for(ConverterFactory factory : converterFactories){
+            if(factory.getTypeIdentifier() == converterFactory.getTypeIdentifier()){
+                return true;
+            }
+        }
+        return false;
+    }
+
+
+    private boolean isClassAlreadyAdded(final ConverterFactory converterFactory) {
```
Comment (hendrikebbers) [inline platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java]: should be renamed:

isAnyConversionTypeAlreadyAdded

File: platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java (hunk 14 above)
Comment (kunsingh) [reply platform/dolphin-platform-remoting-common/src/main/java/com/canoo/dp/impl/remoting/Converters.java]: Done.

**Comment by kunsingh** [comment]
Ready to merge.

**Pull request merged**